llpc/translator/lib/SPIRV/libSPIRV/SPIRVInstruction.h

```diff
@@ -672,6 +680,13 @@ class SPIRVLoad : public SPIRVInstruction, public SPIRVMemoryAccess {
 
   SPIRVValue *getSrc() const { return Module->get<SPIRVValue>(PtrId); }
 
+  virtual void setNonUniform() {
+    if (!hasDecorate(DecorationNonUniform)) {
+      addDecorate(new SPIRVDecorate(DecorationNonUniform, this));
+      getSrc()->setNonUniform();
+    }
+  }
```

Comment:
Why are you propagating through an `OpLoad`?

Consider this code:
```
uint idx = ssbo0.a[foo];
uint loaded = ssbo_array[nonUniformEXT(idx)].b;
```
Only the load from ssbo_array uses non-uniform descriptor indexing. The load from ssbo0 doesn't use non-uniform descriptor indexing, so the OpLoad that is used to load idx should *not* be marked NonUniform.

Reply:
> We probably need to take a step back for this, as this change seems to conflate a few things.
> 
> First, can you explain in detail what fails without your change? The main support for non-uniform image resource access comes from inserting a waterfall loop, which should be keyed off the image instruction, not the descriptor load.
> 
> OpTypeImage/SampledImage cannot go through OpPhi / OpSelect, so it's unclear to me why you'd want to propagate through that in a change that is supposedly about supporting NonUniform for sampled images.

Sorry for not give a clear description on the issue.

**The failed CTS tests like this:**
#version 450
#extension GL_EXT_nonuniform_qualifier : require
#extension GL_EXT_texture_buffer : require
layout(set=0,binding=5) uniform sampler samp[1];
layout(set=0,binding=6) uniform texture2D data[];
layout(location = 0) out vec4     FragColor;
layout(location = 0) in flat int  rIndex;
void main()
{
  // The usage in the spec
  FragColor = texture(nonuniformEXT(sampler2D(data[rIndex], samp[0])), vec2(0,0));
  // The usage not follow spec
  FragColor = texture(sampler2D(data[nonuniformEXT(rIndex)], samp[0]), vec2(0,0)); 
}

**The spir-v text of spec usage:**
OpDecorate %33 NonUniform
...
%4 = OpFunction %void None %3
%5 = OpLabel
%17 = OpLoad %int %16
%19 = OpAccessChain %_ptr_UniformConstant_10 %13 %17
%20 = OpLoad %10 %19
%29 = OpAccessChain %_ptr_UniformConstant_21 %26 %int_0
%30 = OpLoad %21 %29
%32 = OpSampledImage %31 %20 %30
%33 = OpCopyObject %31 %32
%37 = OpImageSampleImplicitLod %v4float %33 %36
    OpStore %9 %37
    OpReturn
    OpFunctionEnd

**The spir-v text of not spec usage:**
OpDecorate %19 NonUniform
OpDecorate %21 NonUniform
OpDecorate %22 NonUniform
...
%main = OpFunction %void None %4
%6 = OpLabel
%18 = OpLoad %int %rIndex
%19 = OpCopyObject %int %18
%21 = OpAccessChain %_ptr_UniformConstant_11 %data %19
%22 = OpLoad %11 %21
%31 = OpAccessChain %_ptr_UniformConstant_23 %samp %int_0
%32 = OpLoad %23 %31
%34 = OpSampledImage %33 %22 %32
%38 = OpImageSampleImplicitLod %v4float %34 %37
    OpStore %FragColor %38
    OpReturn
    OpFunctionEnd

**Root cause:**
The case will failed because the descriptor indexing marked the isNonUniform as false which should be true.
SPIRVToLLVM::transOpAccessChainForImage() {
  ...
  bool isNonUniform = spvIndices[0]->hasDecorate(DecorationNonUniformEXT);
  ...
  return indexDescPtr(base, index, isNonUniform, spvElementType);
}

We need to support the two kinds of usages. So I try to propagate NonUniform form down to up to give indexDescPtr() a right flag for "isNonUniform".
I have a question on OpPhi/OpSelect. I think we can choose a pointer from OpPhi/OpSelect which will be used as the base of OpAccessChain, and the resulting OpAccessChain can be operand of OpSampledImage, is it right?

Reply:
> Why are you propagating through an `OpLoad`?
> 
> Consider this code:
> 
> ```
> uint idx = ssbo0.a[foo];
> uint loaded = ssbo_array[nonUniformEXT(idx)].b;
> ```
> 
> Only the load from ssbo_array uses non-uniform descriptor indexing. The load from ssbo0 doesn't use non-uniform descriptor indexing, so the OpLoad that is used to load idx should _not_ be marked NonUniform.

Yes. You are right. It should be only mark the OpLoad that is used to load OpTypeSampledImage, right?